scriptworker/gpg.py

```diff
@@ -1113,25 +1131,337 @@ def rebuild_gpg_home_signed(context, real_gpg_home, my_pub_key_path,
+    messages = []
+    lines = output.splitlines()
+    status = False
+    for line in lines:
+        m = GOOD.match(line)
+        if m is not None:
+            status = True
+            continue
+        for k, v in BAD.items():
+            if line.startswith(k):
+                messages.append(v)
+                continue
+    if not status:
+        messages.append("No trusted signature!")
+    if messages:
+        raise ScriptWorkerGPGException("\n".join(messages + ["output:", output]))
+
+
+async def get_git_revision(path, exec_function=asyncio.create_subprocess_exec):
+    """Get the git revision of path.
+
+    Args:
+        path (str): the path to run `git rev-parse HEAD` in.
+
+    Returns:
+        str: the revision.
+
+    Raises:
+        ScriptWorkerRetryException: on failure.
+    """
+    proc = await exec_function(
+        'git', "rev-parse", "HEAD", cwd=path,
+        stdout=PIPE, stderr=DEVNULL, stdin=DEVNULL, close_fds=True,
+    )
+    revision, err = await proc.communicate()
+    exitcode = await proc.wait()
+    if exitcode:
+        raise ScriptWorkerRetryException(
+            "Can't get repo revision at {}: {}!".format(path, err)
+        )
+    return revision.decode('utf-8').rstrip()
+
+
+async def update_signed_git_repo(context, repo="origin", revision="master",
+                                 exec_function=asyncio.create_subprocess_exec,
+                                 log_function=pipe_to_log):
+    """Update a git repo with signed git commits, and verify the signature.
+
+    This function updates the repo.
+
+    Args:
+        context (scriptworker.context.Context): the scriptworker context.
+        repo (str, optional): the repo to update from.  Defaults to 'origin'.
+        revision (str, optional): the revision to update to.  Defaults to 'master'.
+
+    Returns:
+        bool: True if there has been a change; False otherwise.
+
+    Raises:
+        ScriptWorkerGPGException: on signature validation failure.
+        ScriptWorkerRetryException: on `git pull` failure.
+    """
+    path = context.config['git_key_repo_dir']
+    old_revision = await get_git_revision(path)
+    proc = await exec_function(
+        "git", "pull", "--ff-only", repo, revision, cwd=path,
+        stdout=PIPE, stderr=STDOUT, stdin=DEVNULL, close_fds=True,
+    )
+    await log_function(proc.stdout)
+    exitcode = await proc.wait()
+    if exitcode:
+        raise ScriptWorkerRetryException(
+            "Can't update repo at {}!".format(path)
+        )
+    revision = await get_git_revision(path)
+    return old_revision != revision
+
+
+async def verify_signed_git_commit(context, path=None,
+                                   exec_function=asyncio.create_subprocess_exec):
+    """Verify `context.config['git_key_repo_dir']` is on a valid signed commit.
+
+    This function calls `verify_signed_git_commit_output` to make sure the
+    latest non-merge-commit commit is signed with a key that lives in
```

Comment:
Nope.  As it stands, we're only checking for the latest non-merge commit being signed.

This means several things:

* signing a commit means you're signing all the previous unsigned commits, which is a little worrisome, but this should be documented.
* however, this allows for creation of pull requests from automation to add pubkeys for taskcluster workers, and then a human only needs to push a final signed commit to the PR to validate it.  I think this will be easier in terms of process, though we could force a rebase here.